File: src/harnesslab/api/routes/local_configuration.py
from __future__ import annotations

import os
from typing import Annotated

from fastapi import APIRouter, Depends, Path, Request, Response
from sqlalchemy.ext.asyncio import AsyncSession

from harnesslab.api.local_configuration import configuration_enabled, local_operator
from harnesslab.api.workbench_dependencies import workbench_session
from harnesslab.api.workbench_errors import WorkbenchAPIError
from harnesslab.productization.assets import distribution_root
from harnesslab.registry import local_harnesses
from harnesslab.registry.connections import (
    ConnectionMetadata,
    ConnectionWrite,
    CredentialMetadata,
    CredentialWrite,
    credential_records,
    list_connections,
    public_credential,
    save_connection,
    save_credential,
)
from harnesslab.registry.local_models import (
    LocalModelConfiguration,
    LocalModelWrite,
    configurations,
    effective_catalog,
    generation_options,
    save_configuration,
)
from harnesslab.registry.service import registry_catalog
from harnesslab.registry.vault import CredentialVault

router = APIRouter(prefix="/local-configuration", tags=["local-configuration"])
Session = Annotated[AsyncSession, Depends(workbench_session)]


@router.get("/status")
async def status(request: Request, response: Response) -> dict[str, bool]:
    response.headers["Cache-Control"] = "no-store"
    return {"enabled": configuration_enabled(request)}


@router.get("/models", dependencies=[Depends(local_operator)])
async def models(
    session: Session, response: Response
) -> dict[str, tuple[LocalModelConfiguration, ...]]:
    response.headers["Cache-Control"] = "no-store"
    return {"items": await configurations(session)}


@router.put("/models/{configuration_id}", dependencies=[Depends(local_operator)])
async def save_model(
    configuration_id: Annotated[str, Path(pattern=r"^[a-z][a-z0-9-]{0,39}$")],
    payload: LocalModelWrite,
    session: Session,
    response: Response,
) -> LocalModelConfiguration:
    response.headers["Cache-Control"] = "no-store"
    return await save_configuration(
        session,
        registry_catalog(distribution_root(), os.environ),
        configuration_id,
        payload,
        await list_connections(session, os.environ, CredentialVault.from_environment()),
    )


@router.get("/credentials", dependencies=[Depends(local_operator)])
async def credentials(
    session: Session, response: Response
) -> dict[str, tuple[CredentialMetadata, ...]]:
    response.headers["Cache-Control"] = "no-store"
    vault = CredentialVault.from_environment()
    return {
        "items": tuple(
            public_credential(row, vault)
            for _, row in sorted((await credential_records(session)).items())
        )
    }


@router.put("/credentials/{credential_id}", dependencies=[Depends(local_operator)])
async def write_credential(
    credential_id: Annotated[str, Path(pattern=r"^[a-z][a-z0-9-]{0,31}$")],
    payload: CredentialWrite,
    session: Session,
    response: Response,
) -> CredentialMetadata:
    response.headers["Cache-Control"] = "no-store"
    return await save_credential(
        session, credential_id, payload, CredentialVault.from_environment()
    )


@router.get("/connections", dependencies=[Depends(local_operator)])
async def connections(
    session: Session, response: Response
) -> dict[str, tuple[ConnectionMetadata, ...]]:
    response.headers["Cache-Control"] = "no-store"
    return {
        "items": await list_connections(session, os.environ, CredentialVault.from_environment())
    }


@router.put("/connections/{connection_id}", dependencies=[Depends(local_operator)])
async def write_connection(
    connection_id: Annotated[str, Path(pattern=r"^[a-z][a-z0-9-]{0,31}$")],
    payload: ConnectionWrite,
    session: Session,
    response: Response,
) -> ConnectionMetadata:
    response.headers["Cache-Control"] = "no-store"
    return await save_connection(
        session,
        registry_catalog(distribution_root(), os.environ),
        connection_id,
        payload,
        os.environ,
        CredentialVault.from_environment(),
    )


@router.post("/connections/{connection_id}/check", dependencies=[Depends(local_operator)])
async def check_connection(
    connection_id: str, session: Session, response: Response
) -> dict[str, object]:
    response.headers["Cache-Control"] = "no-store"
    item = next(
        (
            c
            for c in await list_connections(session, os.environ, CredentialVault.from_environment())
            if c.connection_id == connection_id
        ),
        None,
    )
    if item is None:
        raise WorkbenchAPIError(404, "CONNECTION_NOT_FOUND", "Connection does not exist.")
    return {
        "connection": item,
        "check_kind": "LOCAL_CONFIGURATION_ONLY",
        "connection_health": "NOT_VERIFIED",
        "network_requests": 0,
        "provider_requests": 0,
    }


@router.get("/options", dependencies=[Depends(local_operator)])
async def options(session: Session, response: Response) -> dict[str, object]:
    response.headers["Cache-Control"] = "no-store"
    builtin = registry_catalog(distribution_root(), os.environ)
    models = await configurations(session)
    catalog = effective_catalog(
        builtin,
        models,
        await list_connections(session, os.environ, CredentialVault.from_environment()),
    )
    direct = next(h for h in builtin.harnesses if h.harness_id == "direct-model")
    supported = {p for h in direct.profiles for p in h.supported_provider_profile_ids}
    return {
        "model_controls": [
            generation_options(builtin, p)
            for p in builtin.provider_model_profiles
            if p.profile_id in supported
        ],
        "harness_templates": [
            {
                "profile_id": p.profile_id,
                "harness_id": h.harness_id,
                "version": h.version,
                "profile_reference": p.profile_reference,
                "reasoning_effort": None if h.harness_id == "direct-model" else p.reasoning_effort,
                "tool_surface": h.tool_surface,
                "supported_provider_profile_ids": local_harnesses.allowed_profiles(
                    h, p, catalog, models
                ),
            }
            for h in builtin.harnesses
            for p in h.profiles
        ],
    }


@router.get("/harnesses", dependencies=[Depends(local_operator)])
async def harnesses(
    session: Session, response: Response
) -> dict[str, tuple[local_harnesses.LocalHarnessConfiguration, ...]]:
    response.headers["Cache-Control"] = "no-store"
    return {"items": await local_harnesses.configurations(session)}


@router.put("/harnesses/{configuration_id}", dependencies=[Depends(local_operator)])
async def save_harness(
    configuration_id: Annotated[str, Path(pattern=r"^[a-z][a-z0-9-]{0,39}$")],
    payload: local_harnesses.LocalHarnessWrite,
    session: Session,
    response: Response,
) -> local_harnesses.LocalHarnessConfiguration:
    response.headers["Cache-Control"] = "no-store"
    builtin = registry_catalog(distribution_root(), os.environ)
    models = await configurations(session)
    catalog = effective_catalog(
        builtin,
        models,
        await list_connections(session, os.environ, CredentialVault.from_environment()),
    )
    return await local_harnesses.save_configuration(
        session, builtin, catalog, configuration_id, payload, models
    )

File: frontend/src/api/localConfiguration.ts
import axios from 'axios'
import type { ProviderModelProfile } from '@/types/registry'

export interface LocalModelWrite {
  name: string
  template_profile_id: string
  request_timeout_seconds: number
  enabled: boolean
  expected_revision: number
  purpose?: 'SUBJECT' | 'ANALYST' | 'JUDGE'
  max_output_tokens?: number | null
  reasoning_effort?: string | null
  temperature?: number | null
  connection_id?: string | null
  connection_revision?: number | null
}
export interface CredentialMetadata {
  credential_id: string; revision: number; name: string; enabled: boolean; reference: string; present: boolean
}
export interface ConnectionMetadata {
  connection_id: string; revision: number; name: string; enabled: boolean
  template_provider_id: string; protocol: string; credential_id: string | null; credential_revision: number | null
  environment_reference: string | null; credential_reference: string; endpoint_reference: string; endpoint_fingerprint: string
  endpoint_present: boolean; credential_present: boolean; ready_for_planning: boolean; health_status: 'NOT_VERIFIED'
}
export interface CredentialWrite {
  name: string; expected_revision: number; enabled: boolean; value?: string
}
export interface ConnectionWrite {
  name: string; expected_revision: number; enabled: boolean; template_provider_id: string; protocol: string
  base_url?: string; credential_id: string | null; credential_revision: number | null; environment_reference: string | null
}
export interface LocalModelConfiguration extends Omit<LocalModelWrite, 'expected_revision'> {
  configuration_id: string
  revision: number
  profile: ProviderModelProfile
  template_digest: string
  origin: 'LOCAL'
}
export interface ConfigurationOptions {
  model_controls: { profile_id: string; max_output_tokens: number; reasoning_efforts: string[]; temperature_supported: boolean }[]
  harness_templates: { profile_id: string; harness_id: string; version: string; profile_reference: string; reasoning_effort: string | null; tool_surface: string[]; supported_provider_profile_ids: string[] }[]
}
export interface LocalHarnessWrite {
  name: string; expected_revision: number; enabled: boolean; template_profile_id: string; provider_profile_ids: string[]
}
export interface LocalHarnessConfiguration {
  configuration_id: string; revision: number; name: string; enabled: boolean; template_profile_id: string; template_digest: string; harness_id: string
  profile: { profile_id: string; supported_provider_profile_ids: string[]; harness_config_identity: string }
}
// Operator tokens are passed per request and never installed in shared defaults or storage.
const client = axios.create({ baseURL: '/api/local-configuration', timeout: 15000 })
const headers = (token: string) => ({ Authorization: `Bearer ${token}` })
export const localConfigurationApi = {
  options: async (token: string) => (await client.get<ConfigurationOptions>('/options', { headers: headers(token) })).data,
  harnesses: async (token: string) => (await client.get<{ items: LocalHarnessConfiguration[] }>('/harnesses', { headers: headers(token) })).data,
  saveHarness: async (id: string, payload: LocalHarnessWrite, token: string) => (await client.put<LocalHarnessConfiguration>(`/harnesses/${encodeURIComponent(id)}`, payload, { headers: headers(token) })).data,

  status: async () => (await client.get<{ enabled: boolean }>('/status')).data,
  list: async (token: string) => (await client.get<{ items: LocalModelConfiguration[] }>('/models', { headers: headers(token) })).data,
  save: async (id: string, payload: LocalModelWrite, token: string) => (await client.put<LocalModelConfiguration>(`/models/${encodeURIComponent(id)}`, payload, { headers: headers(token) })).data,
  credentials: async (token: string) => (await client.get<{ items: CredentialMetadata[] }>('/credentials', { headers: headers(token) })).data,
  connections: async (token: string) => (await client.get<{ items: ConnectionMetadata[] }>('/connections', { headers: headers(token) })).data,
  saveCredential: async (id: string, payload: CredentialWrite, token: string) => (await client.put<CredentialMetadata>(`/credentials/${encodeURIComponent(id)}`, payload, { headers: headers(token) })).data,
  saveConnection: async (id: string, payload: ConnectionWrite, token: string) => (await client.put<ConnectionMetadata>(`/connections/${encodeURIComponent(id)}`, payload, { headers: headers(token) })).data,
  checkConnection: async (id: string, token: string) => (await client.post<{ connection: ConnectionMetadata; connection_health: 'NOT_VERIFIED'; network_requests: number; provider_requests: number }>(`/connections/${encodeURIComponent(id)}/check`, {}, { headers: headers(token) })).data,
}
